magento-research/pwa-studio
# [feature]: Add height to `makeUrl` and enforce crop in middleware if provided

## Description
The product images from sample data are 25:31 ratio.
The placeholder image we use is 4:5.

This is a difference of only a few pixels but it leads to a poor UX where we sometimes have a placeholder image that is too large (ie. on the category page) or we have a UI that "shifts" down/up when we render a placeholder and then replace it with the product image (ie. on the pdp).

I made a change here to the `makeUrl` to accept a `height` param. I also added some logic to our middleware to now enforce `crop` if both height and width are provided.

To understand these changes consider the following. Say we have an image with ratio 25:31 (as is our sample data) and we want to size it. The following now can happen:

a) Using `width=640`, sharp and fastly will scale to `640x796` because they retained the aspect ratio.
b) Using `width=640&height=800`, fastly will scale to `640x800` but sharp will scale to `645x800` because of the `fit: outside` [prop](https://github.com/magento-research/express-sharp/blob/develop/lib/transform.js#L23) used. Now, with these changes sharp will scale and then crop to `cover` the provided height and width.

If a user wants to retain the aspect ratio of the image using either fastly or our express-sharp, they can still just pass `width`.

Read more here:
[sharp resize](https://sharp.pixelplumbing.com/en/stable/api-resize/)
[fastly](https://docs.fastly.com/api/imageopto/width)

## Related Issue





Closes #1359 .

## Verification Steps

1. Go to the category page and notice that the images are sized correctly and there is no overflow from the placeholder!
2. Update your backend url to one of our cloud instances with fastly enabled. I can give you a URL if you need one. Then set `IMAGE_OPTIMIZING_ORIGIN="backend"` in your `.env` and re-run your server.
3. Go back to the category page and images should still be sized correctly.


## How have YOU tested this?



I made a change to the category page `GalleryItem` which is only _one_ implementation of `makeUrl`. Ideally we would just us these changes elsewhere when we identify a problem.

## Screenshots / Screen Captures (if appropriate):
In the BEFORE screenshot notice the grey rectangle below the product images.
BEFORE
[![Image from Gyazo](https://i.gyazo.com/cf52b9b14184d77d5b6f8043db4922f4.png)](https://gyazo.com/cf52b9b14184d77d5b6f8043db4922f4)
AFTER
[![Image from Gyazo](https://i.gyazo.com/ff8a21b3f4c7a55a3f7d04ce62782d9c.png)](https://gyazo.com/ff8a21b3f4c7a55a3f7d04ce62782d9c)
## Proposed Labels for Change Type/Package




## Checklist:
